atoma-p2p/src/service.rs

```diff
@@ -0,0 +1,1078 @@
+    ///
+    /// ```rust,ignore
+    /// use atoma_p2p::{AtomaP2pNode, AtomaP2pNodeConfig};
+    /// use std::sync::Arc;
+    /// use sui_keys::keystore::FileBasedKeystore;
+    ///
+    /// async fn setup_node() -> Result<AtomaP2pNode, AtomaP2pNodeError> {
+    ///     let config = AtomaP2pNodeConfig {
+    ///         heartbeat_interval: std::time::Duration::from_secs(1),
+    ///         idle_connection_timeout: std::time::Duration::from_secs(30),
+    ///         listen_addr: "/ip4/127.0.0.1/tcp/8080".to_string(),
+    ///         seed_nodes: vec![],
+    ///         public_url: "node1.example.com".to_string(),
+    ///         node_small_id: 1,
+    ///     };
+    ///     let keystore = Arc::new(FileBasedKeystore::new(&std::path::PathBuf::from("keys"))?);
+    ///     
+    ///     AtomaP2pNode::start(config, keystore, state_manager_sender, false)
+    /// }
+    /// ```
+    ///
+    /// # Network Architecture
+    ///
+    /// The node uses a layered networking approach:
+    /// 1. Transport Layer: TCP/QUIC with noise encryption
+    /// 2. Protocol Layer:
+    ///    - Gossipsub for message broadcasting
+    /// 3. Application Layer: Custom message handling for node addresses and requests
+    ///
+    /// # Security Considerations
+    ///
+    /// - All network connections are encrypted using the noise protocol
+    /// - Messages are signed using the node's private key
+    /// - Peer connections are authenticated
+    /// - The node validates all incoming messages
+    #[instrument(level = "info", skip_all)]
+    pub fn start(
+        config: AtomaP2pNodeConfig,
+        keystore: Arc<FileBasedKeystore>,
+        state_manager_sender: Sender<StateManagerEvent>,
+        is_client: bool,
+    ) -> Result<Self, AtomaP2pNodeError> {
+        let mut swarm = SwarmBuilder::with_new_identity()
+            .with_tokio()
+            .with_tcp(
+                tcp::Config::default(),
+                noise::Config::new,
+                yamux::Config::default,
+            )
+            .map_err(|e| {
+                error!(
+                    target = "atoma-p2p",
+                    event = "build_swarm",
+                    error = %e,
+                    "Failed to build swarm"
+                );
+                AtomaP2pNodeError::SwarmBuildError(e.to_string())
+            })?
+            .with_quic()
+            .with_behaviour(|key| {
+                // To content-address message, we can take the hash of message and use it as an ID.
+                let message_id_fn = |message: &gossipsub::Message| {
+                    let mut s = DefaultHasher::new();
+                    message.data.hash(&mut s);
+                    gossipsub::MessageId::from(s.finish().to_string())
+                };
+
+                let gossipsub_config = gossipsub::ConfigBuilder::default()
+                    .heartbeat_interval(config.heartbeat_interval)
+                    .validation_mode(gossipsub::ValidationMode::Strict)
```

Comment:
You need to add validate_message here
https://docs.rs/gossipsub/latest/gossipsub/struct.GossipsubConfig.html#method.validate_messages

Reply:
Oh nice one, there it goes